lib/include/hse_ikvdb/ikvdb.h

```diff
@@ -46,6 +46,23 @@ enum hse_mclass;
 struct hse_kvdb_txn {
 };
 
+/**
+ * KVDB open mode behavior:
+ *
+ * Mode             Dirty-WAL    Dirty-cNDB    Writes?    Compact?    Queries?    RO FS/Vol
+ * ----------------------------------------------------------------------------------------
+ * rdonly           EUCLEAN      Mem replay    No         No          Yes         Yes
+ * diag             Ignore       Mem replay    No         No          Yes         Yes
+ * rdonly_replay    Replay       Full replay   No         No          Yes         Error
+ * rw (default)     Replay       Full replay   Yes        Yes         Yes         Error
+ */
+enum kvdb_open_mode {
```

Comment:
I would put this enum in the `kvdb_rparams.h` header.

Reply:
How about this to centralize definitions of the strings:
```

#define KVDB_MODE_RDONLY_STR          "rdonly"
#define KVDB_MODE_DIAG_STR            "diag"
#define KVDB_MODE_RDONLY_REPLAY_STR   "rdonly_replay"
#define KVDB_MODE_RDWR_STR            "rw"

#define KVDB_MODE_LIST_STR \
    KVDB_MODE_RDONLY_STR " " \
    KVDB_MODE_DIAG_STR " " \
    KVDB_MODE_RDONLY_REPLAY_STR " " \
    KVDB_MODE_RDWR_STR
```

Use the list macro in help and error messages, for example:
```
        log_err("Invalid value: %s, must be one of: %s", value, KVDB_MODE_LIST_STR);
```

Reply:
Nice idea :)

Reply:
Good idea, will do!

Reply:
I think the internal uses of the open mode would be more clear if we had functions to map a mode to an ability. For example:
```
bool mode_allows_wal_recovery(mode);
bool mode_allows_wal_writes(mode);
bool mode_allows_puts(mode);
bool mode_allows_compaction(mode);
```
Then we can pass mode up and down the call stack (saving it in objects as needed) and not have to have code like this sprinkled around:
```
rdonly_media = (params->mode == KVDB_MODE_RDONLY || params->mode == KVDB_MODE_DIAG);

self->ikdb_read_only = (params->mode != KVDB_MODE_RDWR);

wal->diag_open = (rp->mode == KVDB_MODE_DIAG);

if (mode == KVDB_MODE_RDONLY_REPLAY || mode == KVDB_MODE_RDWR)
    rc = access(home, R_OK | W_OK | X_OK);
else
    rc = access(home, R_OK | X_OK);
```

Reply:
Those are nice interfaces

Reply:
These interfaces could be handy for ikvdb.

The other layers at or below ikvdb such as cndb, c0sk, kvs, c0, cn, mpool doesn't need the mode object.
The only thing that these layers need to know is whether it's a read-only open or not, so just a "rdonly" flag in the respective objects at these different layers would suffice. The only exception here is wal.

Reply:
But it's not obvious what the `object->rdonly` flag is exactly.  Earlier in this comment you can see at least three different expressions that are used to set various `rdonly`-like flags.  A var/member name that is distinct from the flag values would help -- for example, `wal_sync_allowed` and `wal_recovery_allowed`.

Reply:
The APIs might be overkill.  I admit.  But pick var/member names that leave no wiggle room as to their meaning.

Reply:
I have added new interfaces to keep all the kvdb mode related checking centralized..
Also renamed member names in all the layers except mpool to avoid confusion with the mode names. mpool sits on the FS and having rdonly there aligns with the O_RDONLY flag in the FS syscalls.

@alexttx, please let me know if it looks good now.

Reply:
> I would put this enum in the `kvdb_rparams.h` header.

I have added a new file: kvdb_modes.h

Reply:
Muy bueno